frame/dutch-auction/src/lib.rs

```diff
@@ -1,313 +1,311 @@
+			// not recalculates
+			let passed = T::UnixTime::now().as_secs() - order.added_at;
+			let _limit = order.configuration.price(limit, passed)?;
+			let quote = take.amount.safe_mul(&take.limit)?;
 
-			for r in removed.iter() {
-				Orders::<T>::remove(r);
-			}
+			T::MultiCurrency::reserve(order.order.pair.quote, from_to, quote)?;
+			<Takes<T>>::append(order_id, TakeOf::<T> { from_to: from_to.clone(), take });
 
 			Ok(())
 		}
+	}
 
-		fn intention_updated(
-			order_id: &Self::OrderId,
-			action_event: composable_traits::auction::AuctionExchangeCallback,
-		) -> DispatchResult {
-			Orders::<T>::try_mutate(order_id, |order| match order.state {
-				AuctionState::AuctionStarted => {
-					match action_event {
-						composable_traits::auction::AuctionExchangeCallback::Success => {
-							Orders::<T>::remove(order_id);
-							Self::deposit_event(Event::<T>::AuctionSuccess {
-								order_id: order_id.clone(),
-							});
-						},
-						composable_traits::auction::AuctionExchangeCallback::RetryFail => {
-							order.state = AuctionState::AuctionStarted;
-							Orders::<T>::insert(order_id, order);
-						},
-						composable_traits::auction::AuctionExchangeCallback::FatalFail => {
-							Orders::<T>::remove(order_id);
-							Self::deposit_event(Event::<T>::AuctionFatalFail {
-								order_id: order_id.clone(),
-							});
-						},
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		// this cleanups all takes added into block, so we never store takes
+		// so we stay fast and prevent attack
+		fn on_finalize(_n: T::BlockNumber) {
+			for (order_id, mut takes) in <Takes<T>>::iter() {
+				// users payed N * WEIGHT before, we here pay N * (log N - 1) * Weight. We can
+				// retain pure N by first served principle so, not highest price.
+				takes.sort_by(|a, b| b.take.limit.cmp(&a.take.limit));
+				let SellOrder { mut order, added_at: _, from_to: ref seller, configuration: _ } =
+					<SellOrders<T>>::get(order_id)
+						.expect("takes are added only onto existing orders");
+
+				// calculate real price
+				for take in takes {
+					let take_amount = take.take.amount.min(order.take.amount);
+					order.take.amount -= take_amount;
+					let quote_amount = take_amount.saturating_mul(take.take.limit);
+
+					exchange_reserved::<T>(
+						order.pair.base,
+						seller,
+						take_amount,
+						order.pair.quote,
+						&take.from_to,
+						quote_amount,
+					)
+					.expect("we forced locks beforehand");
+
+					// what to do with orders which nobody ever takes? some kind of dust orders with
+					// 1 token
+					if order.take.amount == T::Balance::zero() {
+						break
 					}
-					Ok(())
-				},
-				_ => Ok(()),
-			})
+				}
+
+				if order.take.amount == T::Balance::zero() {
+					<SellOrders<T>>::remove(order_id);
+					Self::deposit_event(Event::OrderRemoved { order_id });
+				}
+			}
+			<Takes<T>>::remove_all(None);
```

Comment:
Quick question: why not let the user directly execute the take in the extrinsic transaction without having a hook?

Reply:
- hydra and other books work like that
- you can predict price, block and start sending minimal price as soon as it close. will get minimal amount for collateral.
- but if on some step, several people will decide that it worth it, and they will know that highest bid wins, they will be able to pay little more to overbet. price will be more optimal than predicted by auction.
- if to take directly, than bot in same data center as we host will win bets.
- also i have heard transaction tip to be first, may be it can work too. so it will be tip to composable, not to dutch auction or lending. and need to read more how tip works.

Reply:
@hussein-aitlahcen what do you think?

Reply:
makes sense, thanks for clarifying